pycloudlib/vmware/instance.py

```diff
@@ -0,0 +1,94 @@
+# This file is part of pycloudlib. See LICENSE file for license information.
+"""VMWare instance."""
+import subprocess
+from typing import Mapping, Optional
+
+from pycloudlib.errors import PycloudlibTimeoutError
+from pycloudlib.instance import BaseInstance
+from pycloudlib.key import KeyPair
+
+
+class VMWareInstance(BaseInstance):
+    """VMWare backed instance."""
+
+    _type = "vmware"
+
+    def __init__(self, key_pair: KeyPair, vm_id: str, env: Mapping):
+        """Instance for VMWare.
+
+        Args:
+            key_pair: SSH key object
+            vm_id: ID of the VM to interact with
+        """
+        super().__init__(key_pair=key_pair)
+        self.vm_id = vm_id
+        self._ip: Optional[str] = None
+        self.env = env
+
+    @property
+    def name(self) -> str:
+        """Return VM name."""
+        return self.vm_id
+
+    @property
+    def ip(self) -> str:
+        """Return VM IP."""
+        if self._ip is not None:
+            return self._ip
+        ips_output = subprocess.check_output(
+            ["govc", "vm.ip", "-wait", "5m", self.vm_id], env=self.env
+        )
+        if not ips_output:
+            raise PycloudlibTimeoutError(
+                "Could not retrieve IP address after 5 minutes"
+            )
+
+        ips = ips_output.decode().strip().split(",")
+        if len(ips) > 1:
+            self._log.warning("Expected 1 IP but got %s. Using the first", ips)
+        self._ip = ips[0]
+        return ips[0]  # Not self._ip here because mypy isn't smart enough
+
+    def delete(self, wait=True):
+        """Delete the instance."""
+        self.shutdown()
+        subprocess.run(
+            ["govc", "vm.destroy", self.vm_id], env=self.env, check=True
+        )
+
+    def _do_restart(self, **kwargs):
+        """Restart the instance."""
+        self.shutdown()
+        self.start()
+
+    def shutdown(self, wait=True, **kwargs):
+        """Shutdown the instance.
+
+        Args:
+            wait: wait for the instance shutdown
+        """
+        subprocess.run(
+            ["govc", "vm.power", "-off", self.vm_id], env=self.env, check=True
+        )
+
+    def start(self, wait=True):
+        """Start the instance.
+
+        Args:
+            wait: wait for the instance to start.
+        """
+        subprocess.run(
+            ["govc", "vm.power", "-on", self.vm_id], env=self.env, check=True
```

Comment:
Should this command be idempotent? As in, does it exit non-zero of power is already on? If so maybe we should `pass` on that exception because start was already performed? Same question for shutdown, should we succeed if power is already off or do we consider that a failure case as the VM maybe lost power due to something outside of test harness interaction.

Reply:
I think idempotent makes more sense here. Thanks for the idea.